src/compiler/checker.ts

```diff
@@ -30395,6 +30402,9 @@ namespace ts {
                     assignBindingElementTypes(declaration.name);
                 }
             }
+            else if (type) {
+                Debug.assertEqual(links.type, type);

```

Comment:
This is a good idea, but I think I'd wait until after the beta to merge this assertion.

(based on my experience putting a subtle assertion into overload error reporting)